crates/bevy_ecs/src/schedule/executor/async.rs

```diff
@@ -0,0 +1,432 @@
+use crate::schedule::r#async::keyed_queues::KeyedQueues;
+use crate::schedule::{InternedScheduleLabel, ScheduleLabel};
+use crate::system::RunSystemError;
+use crate::world::unsafe_world_cell::UnsafeWorldCell;
+use crate::world::FromWorld;
+use crate::{
+    system::{SystemParam, SystemState},
+    world::World,
+};
+use bevy_ecs::world::{Mut, WorldId};
+use bevy_platform::collections::HashMap;
+use bevy_platform::sync::{Arc, Mutex, OnceLock, RwLock};
+use concurrent_queue::ConcurrentQueue;
+use core::any::TypeId;
+use core::marker::PhantomData;
+use core::pin::Pin;
+use core::sync::atomic::{AtomicI64, AtomicUsize, Ordering};
+use core::task::{Context, Poll, Waker};
+use std::thread;
+
+mod keyed_queues {
+    use concurrent_queue::ConcurrentQueue;
+    use std::sync::Arc;
+    use std::{collections::HashMap, hash::Hash, sync::RwLock};
+
+    /// HashMap<K, Arc<ConcurrentQueue<V>>> behind a single RwLock.
+    /// - Writers only contend when creating a new key or GC'ing.
+    /// - `push` is non-blocking (unbounded queue).
+    pub struct KeyedQueues<K, V> {
```

Comment:
Can you expand the doc comment to explain what a KeyedQueue is and why you would use it? And maybe move the info relevant to `inner` onto `inner`.